### PR TITLE
JBTM-2982 Drive LRA participants from a separate thread

### PR DESCRIPTION
**Pull request by mmusgrov**
https://issues.jboss.org/browse/JBTM-2982

!TOMCAT !RTS !QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF !AS_TESTS !mysql !postgres !db2 !oracle

### CODE REVIEW

File: rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java [hunk 1]
```diff
@@ -379,15 +371,19 @@ private int tryDoEnd(boolean compensate) {
...
+        updateStatus(compensate);
 
         // if the the request is still in progress (ie accepted is true) let recovery finish it
         return accepted ? TwoPhaseOutcome.HEURISTIC_HAZARD : TwoPhaseOutcome.FINISH_OK;
     }
 
+    private void updateStatus(boolean compensate) {
+           if (compensate)
+            status = accepted ? CompensatorStatus.Compensating : CompensatorStatus.Compensated;
+        else
```
Comment (ochaloup) [inline rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java]: maybe fixing indentation?

File: rts/lra/lra-test/pom.xml [hunk 2]
```diff
@@ -35,14 +35,14 @@
              -Xrunjdwp:transport=dt_socket,address=${coordinator.debug.port},server=y,suspend=y
              -->
         <coordinator.debug.port>8787</coordinator.debug.port>
-        <coordinator.debug.jdwp>-Xrunjdwp:transport=dt_socket,address=${coordinator.debug.port},server=y,suspend=y</coordinator.debug.jdwp>
+        <coordinator.debug.jdwp>-Xrunjdwp:transport=dt_socket,address=${coordinator.debug.port},server=y,suspend=n</coordinator.debug.jdwp>
         <coordinator.debug.params></coordinator.debug.params>
 
         <!-- config for the test JAX-RS resource, for exampe to enable debugging use
```
Comment (ochaloup) [inline rts/lra/lra-test/pom.xml]: do we need here this comment when the `swarm.debug.jdwp` attribute covers this?

File: rts/lra/lra-test/pom.xml [hunk 3]
```diff
@@ -35,14 +35,14 @@
              -Xrunjdwp:transport=dt_socket,address=${coordinator.debug.port},server=y,suspend=y
```
Comment (ochaloup) [inline rts/lra/lra-test/pom.xml]: do we need here this comment when the swarm.debug.jdwp attribute covers this?

File: rts/lra/lra-test/src/main/java/io/narayana/lra/participant/api/ActivityController.java [hunk 4]
```diff
@@ -560,8 +560,25 @@ private void endCheck(Activity activity) {
...
+                } else {
+                    int ms = 0;
+
+                    try {
+                        ms = Integer.getInteger(arg, 0);
+                    } catch (Exception ignore) {
+                    }
+
+                    try {
+                        Thread.sleep(ms <= 0 ? Long.MAX_VALUE : ms); // delay the end call
```
Comment (ochaloup) [inline rts/lra/lra-test/src/main/java/io/narayana/lra/participant/api/ActivityController.java]: I would like to understand the rationalization of giving to user a way of delaying end call?
It's seems as some workaround to something. I can be a wrong in understanding but letting the client to drive timing from outside sounds to me supportive of race conditions.
I know we discussed some timeouts to be involved but I can't connect this with the discussion.

File: rts/lra/lra-test/pom.xml (hunk 2 above)
Comment (mmusgrov) [reply rts/lra/lra-test/pom.xml]: What do you mean, do you want me to add an explanation of how the debugging config works?

File: rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java (hunk 1 above)
Comment (mmusgrov) [reply rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java]: done

File: rts/lra/lra-test/pom.xml (hunk 2 above)
Comment (ochaloup) [reply rts/lra/lra-test/pom.xml]: sure. I'm sorry for not being clear. I just meant that if there is declared `coordinator.debug.jdwp` with debug parameters the comment could be removed completely. The information is doubled.

File: rts/lra/lra-test/pom.xml (hunk 3 above)
Comment (ochaloup) [reply rts/lra/lra-test/pom.xml]: I just meant if information is not doubled and if you consider so the comment could be removed competely.

**Comment by ochaloup** [comment]
@mmusgrov hm... I have probably different version, I can't find the `isRecovering` at more places in the `LRARecord`.  If I take what is here in the PR: https://github.com/mmusgrov/narayana/blob/43e5c9c9a93cf58db1b25b8a1a86b76e27622bd2/rts/lra/lra-coordinator/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java#L242 then the variable is declared at the `l242` and then I can't find any usage of it. What is your version of the code?

File: rts/lra/lra-test/src/main/java/io/narayana/lra/participant/api/ActivityController.java (hunk 4 above)
Comment (ochaloup) [reply rts/lra/lra-test/src/main/java/io/narayana/lra/participant/api/ActivityController.java]: ouch, I see. Even I was checking the package I missed that. I'm sorry,  this is fine.

File: rts/lra/lra-test/pom.xml (hunk 3 above)
Comment (mmusgrov) [reply rts/lra/lra-test/pom.xml]: okay

**Comment by mmusgrov** [comment]
@ochaloup re *isRecovering*, it was me that was looking at a wrong version, sorry.

So I have pushed a commit with that, the extra tab and the jdwp debug comment removed. None of those are code changes so if you are satisfied then I don't think we need to wait for CI?

File: rts/lra/lra-test/src/main/java/io/narayana/lra/participant/api/ActivityController.java (hunk 4 above)
Comment (mmusgrov) [reply rts/lra/lra-test/src/main/java/io/narayana/lra/participant/api/ActivityController.java]: Notice that I also added a comment to indicate what the recovery scan is testing here.

**Pull request merged**